kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,995 @@
+                    return Err(Error::Generic(format!(
+                        "Requested field not found in parquet schema, and field is not nullable: {}",
+                        field.name()
+                    )));
+                }
+            }
+        }
+    }
+    Ok((
+        parquet_offset + fields.len() - start_parquet_offset,
+        reorder_indices,
+    ))
+}
+
+/// Get the indices in `parquet_schema` of the specified columns in `requested_schema`. This returns
+/// a tuples of (mask_indices: Vec<parquet_schema_index>, reorder_indices:
+/// Vec<requested_index>). `mask_indices` is used for generating the mask for reading from the
 /// parquet file, and simply contains an entry for each index we wish to select from the parquet
-/// file set to the index of the requested column in the parquet. `reorder_indicies` is used for
-/// re-ordering and will be the same size as `requested_schema`. Each index in `reorder_indicies`
-/// represents a column that will be in the read parquet data at that index. The value stored in
-/// `reorder_indicies` is the position that the column should appear in the final output. For
-/// example, if `reorder_indicies` is `[2,0,1]`, then the re-ordering code should take the third
-/// column in the raw-read parquet data, and move it to the first column in the final output, the
-/// first column to the second, and the second to the third.
+/// file set to the index of the requested column in the parquet. `reorder_indices` is used for
+/// re-ordering. See the documentation for [`ReorderIndex`] to understand what each element in the
+/// returned array means
 pub(crate) fn get_requested_indices(
     requested_schema: &SchemaRef,
     parquet_schema: &ArrowSchemaRef,
-) -> DeltaResult<(Vec<usize>, Vec<usize>)> {
-    let (mask_indicies, reorder_indicies): (Vec<usize>, Vec<usize>) = parquet_schema
-        .fields()
-        .iter()
-        .enumerate()
-        .filter_map(|(parquet_index, field)| {
-            requested_schema
-                .index_of(field.name())
-                .map(|index| (parquet_index, index))
-        })
-        .unzip();
-    require!(
-        mask_indicies.len() == requested_schema.fields.len(),
-        Error::generic("Didn't find all requested columns in parquet schema")
-    );
-    Ok((mask_indicies, reorder_indicies))
+) -> DeltaResult<(Vec<usize>, Vec<ReorderIndex>)> {
+    let mut mask_indices = vec![];
+    let (_, reorder_indexes) = get_indices(
+        0,
+        requested_schema,
+        parquet_schema.fields(),
+        &mut mask_indices,
+    )?;
+    Ok((mask_indices, reorder_indexes))
 }
 
-/// Create a mask that will only select the specified indicies from the parquet. Currently we only
+/// Create a mask that will only select the specified indices from the parquet. Currently we only
 /// handle "root" level columns, and hence use `ProjectionMask::roots`, but will support leaf
 /// selection in the future. See issues #86 and #96 as well.
```

Comment:
Yep, I did actually update the comment, but I guess not before you looked at this :)

For the args, I've kept them in because of the TODO. I'd like to run some tests to see if there's any different between not setting a `ProjectionMask` vs. a `leaves` mask that has all the fields.

If it's different, I'd add an optimization here to detect that and return `None`, if there's no real difference, I'll remove the args and the optional part of the return type. I just figured I wouldn't bother to make all the changes to the call sites until I was sure which way to go.